src/main/java/com/warrenstrange/googleauth/GoogleAuthenticator.java

```diff
@@ -582,24 +588,27 @@ private ICredentialRepository getValidCredentialRepository()
         return repository;
     }
 
-    /**
-     * This method loads the first available ICredentialRepository
-     * registered using the Java service loader API.
-     *
-     * @return the first registered ICredentialRepository or <code>null</code>
-     * if none is found.
-     */
-    private ICredentialRepository getCredentialRepository()
+    @Override
+    public ICredentialRepository getCredentialRepository()
     {
-        ServiceLoader<ICredentialRepository> loader =
-                ServiceLoader.load(ICredentialRepository.class);
+        if (credentialRepository == null) {
```

Comment:
You are using an `AtomicReference`, suggesting you're providing thread-safe operations to deal with the credential repository, yet there's a memory visibility problem here: there is no guarantee you see changes to `credentialRepository` from another thread executing `getCredentialRepository` or `setCredentialRepository`).  A proper way to do this would be creating the `AtomicReference` instance earlier on and then use `compareAndSet`.  Perhaps, it's just easier _not_ to use atomic classes at all in this case.